### PR TITLE
Self-distillation: 15min 2-layer teacher, 15min 1-layer student

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
Train a 2-layer teacher (n_hidden=128) for 15 min, then distill into the 1-layer student for 15 min. The teacher's superior representations (especially for tandem) transfer to the student's faster architecture. Combines depth benefit with throughput.

## Instructions
This is a complex experiment with two phases:

1. Phase 1 (15 min): Train 2-layer model:
\`\`\`python
teacher_config = dict(n_hidden=128, n_layers=2, n_head=4, slice_num=48, ...)
\`\`\`
Save checkpoint at 15 min mark.

2. Phase 2 (15 min): Load 1-layer student (current architecture), add KD loss:
\`\`\`python
with torch.no_grad():
    teacher_pred = teacher_model(batch)["preds"].float()
kd_loss = (student_pred - teacher_pred.detach()).abs().mean()
loss = loss + 0.5 * kd_loss  # blend with original L1 loss
\`\`\`

3. Keep original loss active throughout Phase 2 (KD is auxiliary).

Run with \`--wandb_group self-distill-2to1\`.

## Baseline
val_loss=0.8495 | in=17.84 | ood_c=13.66 | ood_r=27.77 | tan=36.36

---
## Results

**W&B run:** \`u8zz8l07\`
**Total wall time:** 30.6 min (15.1 min teacher + 15.4 min student)
**Peak memory:** 20.1 GB
**Teacher:** 27 epochs | **Student:** 21 epochs

### Student metrics (best at epoch 21)

| Split | val_loss | mae_surf_Ux | mae_surf_Uy | mae_surf_p | mae_vol_p |
|---|---|---|---|---|---|
| val_in_dist | 1.8044 | 11.88 | 3.36 | 59.07 | 55.64 |
| val_tandem_transfer | 2.6434 | 8.83 | 3.25 | 62.40 | 59.57 |
| val_ood_cond | 1.9090 | 8.34 | 2.19 | 41.15 | 32.69 |
| val_ood_re | 1.4007 | 8.11 | 2.04 | 46.22 | 61.71 |
| **combined val/loss** | **1.9394** | | | | |

**mean3 surf_p** = (59.1 + 62.4 + 41.2) / 3 = **54.2** vs baseline **22.6** — 2.4× worse

### Comparison vs baseline

| Metric | Baseline | This run | Δ |
|---|---|---|---|
| val/loss | 0.8495 | 1.9394 | +128% |
| surf_p in_dist | 17.84 | 59.07 | +231% |
| surf_p tandem | 36.36 | 62.40 | +72% |
| surf_p ood_cond | 13.66 | 41.15 | +201% |
| surf_p ood_re | 27.77 | 46.22 | +66% |

### What happened

**The approach failed decisively.** The 30-minute budget split into two 15-minute phases means neither teacher nor student reaches adequate convergence:

1. **Underfitted teacher**: Phase 1 runs 27 epochs. EMA never activates (starts at epoch 40). The teacher val/loss peaks at ~1.3 per split vs baseline ~0.85. Soft targets from an underfitted teacher add noise rather than signal.

2. **Student gets fewer epochs with more overhead**: The teacher forward pass in Phase 2 adds ~35% per-epoch cost (43s vs 32s baseline). The student only reaches 21 epochs in 15 min — vs 52 epochs in the standard 30-min run.

3. **Same failure mode as model soup**: Splitting the 30-min budget into sub-runs creates underfitted components that don't combine beneficially. Full 30 min on a single run always wins.

4. **KD actually hurts convergence speed**: Even if the teacher were perfect, the teacher inference overhead directly eats into the student's training budget.

### Suggested follow-ups

1. **Pretrained teacher (no extra budget)**: Use an existing well-trained 2-layer checkpoint from a prior run (e.g., `dh7o63lr`, the first depth-iso-flops run) as a fixed teacher. Full 30 min for student with KD. This is the version most likely to work.
2. **Warm-start student from teacher**: Initialize student weights from the teacher checkpoint (layer 0 only), then fine-tune for the full 30 min without KD overhead.
3. **Feature distillation**: Distill intermediate attention maps instead of output predictions to reduce per-batch overhead.

### CODE REVIEW

**Comment by tcapelle** [comment]
Closing — self-distillation failed catastrophically (val_loss=1.9394). The two-phase training approach was too complex for the 30-min budget.
